### PR TITLE
pkg/gadgets/trace: remove cgo

### PR DESCRIPTION
**Pull request by flyth**
This PR removes dependencies on cgo for all trace gadgets.

### CODE REVIEW

File: pkg/gadgets/trace/tcp/tracer/tracer.go [hunk 1]
```diff
@@ -222,41 +190,37 @@ func (t *Tracer) run() {
...
-		if eventC._type == C.TCP_EVENT_TYPE_CONNECT {
+		if eventC.Type == tcptracerEventTypeTCP_EVENT_TYPE_CONNECT {
 			event.Operation = "connect"
-		} else if eventC._type == C.TCP_EVENT_TYPE_ACCEPT {
+		} else if eventC.Type == tcptracerEventTypeTCP_EVENT_TYPE_ACCEPT {
 			event.Operation = "accept"
-		} else if eventC._type == C.TCP_EVENT_TYPE_CLOSE {
+		} else if eventC.Type == tcptracerEventTypeTCP_EVENT_TYPE_CLOSE {
 			event.Operation = "close"
 		}
```
Comment (mauriciovasquezbernal) [inline pkg/gadgets/trace/tcp/tracer/tracer.go:222]: I know it's not related to this PR, but can we take the opportunity to replace by a switch statement?

File: pkg/gadgets/trace/tcpconnect/tracer/bpf/tcpconnect.h [hunk 2]
```diff
@@ -25,14 +25,16 @@ struct ipv6_flow_key {
 
 struct event {
 	union {
+		unsigned __int128 saddr;
 		__u32 saddr_v4;
 		__u8 saddr_v6[16];
```
Comment (mauriciovasquezbernal) [inline pkg/gadgets/trace/tcpconnect/tracer/bpf/tcpconnect.h:30]: I suppose you can remove it and `daddr_v6` and apply the following diff:

```
diff --git pkg/gadgets/trace/tcpconnect/tracer/bpf/tcpconnect.bpf.c pkg/gadgets/trace/tcpconnect/tracer/bpf/tcpconnect.bpf.c
index a8e835a9..de1f0b02 100644
--- pkg/gadgets/trace/tcpconnect/tracer/bpf/tcpconnect.bpf.c
+++ pkg/gadgets/trace/tcpconnect/tracer/bpf/tcpconnect.bpf.c
@@ -156,9 +156,9 @@ trace_v6(struct pt_regs *ctx, pid_t pid, struct sock *sk, __u16 dport, __u64 mnt
        event.uid = bpf_get_current_uid_gid();
        event.ts_us = bpf_ktime_get_ns() / 1000;
        event.mntns_id = mntns_id;
-       BPF_CORE_READ_INTO(&event.saddr_v6, sk,
+       BPF_CORE_READ_INTO(&event.saddr, sk,
                           __sk_common.skc_v6_rcv_saddr.in6_u.u6_addr32);
-       BPF_CORE_READ_INTO(&event.daddr_v6, sk,
+       BPF_CORE_READ_INTO(&event.daddr, sk,
                           __sk_common.skc_v6_daddr.in6_u.u6_addr32);
        event.dport = dport;
        bpf_get_current_comm(event.task, sizeof(event.task));
```

**Review by mauriciovasquezbernal — COMMENTED** [review]
It's looking great. Some small comments around.

File: pkg/gadgets/trace/bind/tracer/bpf/bindsnoop.bpf.c [hunk 3]
```diff
@@ -15,6 +15,8 @@ const volatile bool ignore_errors = true;
 const volatile bool filter_by_port = false;
 const volatile bool filter_by_mnt_ns = false;
 
+const struct bind_event *unusedbindevent __attribute__((unused));
+
```
[the remark this answers is not in the document]
Comment (flyth) [reply pkg/gadgets/trace/bind/tracer/bpf/bindsnoop.bpf.c:19]: Not having this makes the compiler "optimize" it away, see [here](https://github.com/cilium/ebpf/discussions/627).

File: pkg/gadgets/trace/bind/tracer/tracer.go [hunk 4]
```diff
@@ -267,13 +244,10 @@ func (t *Tracer) run() {
 			continue
 		}
 
-		eventC := (*C.struct_bind_event)(unsafe.Pointer(&record.RawSample[0]))
-
-		addr := C.ip_to_string(eventC)
-		defer C.free(unsafe.Pointer(addr))
+		eventC := (*bindsnoopBindEvent)(unsafe.Pointer(&record.RawSample[0]))
```
[the remark this answers is not in the document]
Comment (flyth) [reply pkg/gadgets/trace/bind/tracer/tracer.go:247]: I like that it's still showing that it's coming from c/ebpf. Maybe to keep it generic call it ebpfEvent / bpfEvent, then? WDYT?

File: pkg/gadgets/trace/dns/tracer/tracer.go [hunk 5]
```diff
@@ -302,36 +270,40 @@ func parseDNSEvent(rawSample []byte) (*types.Event, error) {
...
+	if dnsEvent.Qr == 1 {
 		event.Qr = types.DNSPktTypeResponse
-		srcAddr := C.get_src_addr(dnsEvent)
-		event.Nameserver = C.GoString(srcAddr)
-		C.free(unsafe.Pointer(srcAddr))
+		if dnsEvent.Af == syscall.AF_INET {
+			event.Nameserver = netip.AddrFrom4(*(*[4]byte)(dnsEvent.SaddrV6[0:4])).String()
+		} else if dnsEvent.Af == syscall.AF_INET6 {
+			event.Nameserver = netip.AddrFrom16(dnsEvent.SaddrV6).String()
+		}
```
[the remark this answers is not in the document]
Comment (flyth) [reply pkg/gadgets/trace/dns/tracer/tracer.go:286]: Yeah, that conversion happens a couple of times - however: uint128 isn't (yet) a native type to Go (although cilium uses it as type), so we can't use that globally, yet. However, a solution could be to define the ip addresses in C as unions (in most cases they are already) where the first type leads to a [16]byte and then use that in a helper function. I'll look into it.

File: pkg/gadgets/trace/sni/tracer/tracer.go [hunk 6]
```diff
@@ -135,18 +131,6 @@ func (t *Tracer) Attach(
...
-	copy(name, rawSample)
-
-	str := string(name)
-	i := strings.Index(str, "\x00")
-	if i > 0 {
-		return str[:i]
-	}
-	return str
-}
-
```
[the remark this answers is not in the document]
Comment (flyth) [reply pkg/gadgets/trace/sni/tracer/tracer.go:149]: Haha, yeah, but now that I see it, we have to limit the slice to TLS_MAX_SERVER_NAME_LEN to get the same result in any case.

File: pkg/gadgets/trace/bind/tracer/tracer.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/gadgets/trace/bind/tracer/tracer.go]: I see. Yes, either of your proposals work for me.

File: pkg/gadgets/trace/sni/tracer/tracer.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (flyth) [reply pkg/gadgets/trace/sni/tracer/tracer.go:149]: I haven't seen a way to export consts, but I'll recheck. A way that might work for now would be using an enum, but that seems pretty hacky.

**Review by mauriciovasquezbernal — APPROVED** [review]
IMO it's ready to go. Thanks for this, it helps a lot!

**Comment by mauriciovasquezbernal** [comment]
Are we missing something to merge this? 

(Please notice that there are some `fixup` commits to be squashed before merging)

**Pull request merged**